### PR TITLE
subcall wrapper + unit tests, and staking address generation

### PR DESCRIPTION
**Pull request by CedarMist**
This implements the following runtime-sdk functions:

 * consensus.Undelegate
 * consensus.Delegate
 * consensus.Withdraw
 * accounts.Transfer

It also supports generation of staking wallet addresses. via `Subcall.sol`:`ConsensusUtils.generateStakingAddress`

consensus.Deposit isn't supported.

I will need to update the documentation to include some subcall examples using the Subcall.sol library.

Subcall is no-longer in Sapphire.sol, all the functionality for subcalls is implemented in Subcall.sol.

### CODE REVIEW

File: contracts/contracts/Subcall.sol [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+    /**
+     * Delegate native token to consensus level
+     *
+     * @param to Staking account
+     * @param value native token amount (in wei)
+     */
+    function consensusDelegate(StakingPublicKey to, uint128 value) internal {
+        (uint64 status, bytes memory data) = _subcallWithToAndAmount(
+            "consensus.Delegate",
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:168]: Move to a constant above, so we have all method names in one place.

FYI go, typescript and rust bindings:
- https://github.com/oasisprotocol/oasis-sdk/tree/main/client-sdk/go/modules
- https://github.com/oasisprotocol/oasis-sdk/tree/main/client-sdk/ts-web/rt/src
- https://github.com/oasisprotocol/oasis-sdk/tree/main/runtime-sdk/src/modules

File: contracts/contracts/Subcall.sol [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+    /**
+     * Transfer from an account in this runtime to a consensus staking account.
+     *
+     * @param to consensus staking account which gets the tokens
+     * @param value native token amount (in wei)
+     */
+    function consensusWithdraw(StakingPublicKey to, uint128 value) internal {
+        (uint64 status, bytes memory data) = _subcallWithToAndAmount(
+            "consensus.Withdraw",
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:188]: Move to a constant.

File: contracts/contracts/Subcall.sol [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+     * Perform a transfer to another account
+     *
+     * This is equivalent of `payable(to).transfer(value);`
+     *
+     * @param to Destination account
+     * @param value native token amount (in wei)
+     */
+    function accountsTransfer(address to, uint128 value) internal {
+        (uint64 status, bytes memory data) = _subcallWithToAndAmount(
+            "accounts.Transfer",
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:210]: Move to a constant.

File: contracts/test/subcall.ts [hunk 4]
```diff
@@ -1,28 +1,193 @@
...
+      'this key is not important',
+    );
+    const computedPublicKey = ethers.utils.hexlify(
+      await oasis.staking.addressFromPublicKey(alice.public()),
+    );
+
+    expect(computedPublicKey).eq(ethers.utils.hexlify(newKeypair.publicKey));
+  });
+
+  it('accounts.Transfer', async () => {
```
Comment (matevz) [inline contracts/test/subcall.ts:94]:
```suggestion
  it(oasis.accounts.METHOD_TRANSFER, async () => {
```

File: contracts/test/subcall.ts [hunk 5]
```diff
@@ -1,28 +1,193 @@
...
+
+    // Transfer using the Subcall.accounts_Transfer method
+    tx = await contract.testAccountsTransfer(ownerAddr, 1);
+    receipt = await tx.wait();
+
+    // Ensure contract only no wei left
+    expect(await contract.provider.getBalance(contract.address)).eq(0);
+  });
+
+  it('consensus.Delegate', async () => {
```
Comment (matevz) [inline contracts/test/subcall.ts:127]: Does this exist?
```suggestion
  it(oasis.consensusAccounts.METHOD_DELEGATE, async () => {
```

File: contracts/test/subcall.ts [hunk 6]
```diff
@@ -1,28 +1,193 @@
...
+    await ensureBalance(contract, initialBalance, owner);
+
+    let tx = await contract.testConsensusUndelegate(kp.publicKey, 0);
+    await tx.wait();
+    expect(await contract.provider.getBalance(contract.address)).eq(
+      initialBalance,
+    );
+  });
+
+  it('consensus.Withdraw', async () => {
```
Comment (matevz) [inline contracts/test/subcall.ts:178]:
```suggestion
  it(oasis.consensusAccounts.METHOD_WITHDRAW, async () => {
```

File: contracts/test/subcall.ts [hunk 7]
```diff
@@ -1,28 +1,193 @@
...
+          : new TextDecoder().decode(ethers.utils.arrayify(event.data)),
+    };
+    expect(event.status).eq(0);
+    expect(cborg.decode(ethers.utils.arrayify(event.data))).is.null;
+
+    // Ensure contract only no wei left
+    //expect(await contract.provider.getBalance(contract.address)).eq(0);
+  });
+
+  it('consensus.Undelegate', async () => {
```
Comment (matevz) [inline contracts/test/subcall.ts:166]:
```suggestion
  it(oasis.consensusAccounts.METHOD_UNDELEGATE, async () => {
```

**Review by matevz — CHANGES_REQUESTED** [review]

File: contracts/contracts/Subcall.sol [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+                        )
+                    )
+                )
+            );
+    }
+}
+
+library Subcall {
+    address internal constant SUBCALL =
+        0x0100000000000000000000000000000000000103;
```
[the remark this answers is not in the document]
Comment (CedarMist) [reply contracts/contracts/Subcall.sol:77]: I want to remove `subcall` from the Sapphire.sol library.

There is no need for the two libraries to linked together. So everything subcall related lives in Subcall.sol.

But, then we import Sapphire.sol already for ConsensusUtils

File: contracts/test/subcall.ts (hunk 4 above)
Comment (CedarMist) [reply contracts/test/subcall.ts:94]: The runtime SDK modules are in the `oasisRT` package.

The latest package version on npmjs.com is 0.2.1-alpha.2 (published last year)

This does not include the constants I need.

I don't think it's a good idea to import a whole package just to use some string constants, that's how the JS ecosystem ended up like it did and I'm not going to help perpetuate the enspaghettification.

File: contracts/contracts/Subcall.sol [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        );
+    }
+
+    error ConsensusUndelegateError(uint64 status, string data);
+
+    /**
+     * Start the undelegation process of the given number of shares from
+     * consensus staking account to runtime account.
+     *
+     * @param from Public key which shares were delegated to
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:70]:
```suggestion
     * @param from Consensus address which shares were delegated to
```

File: contracts/contracts/Subcall.sol [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+            revert ConsensusUndelegateError(status, string(data));
+        }
+    }
+
+    error ConsensusDelegateError(uint64 status, string data);
+
+    /**
+     * Delegate native token to consensus level
+     *
+     * @param to Staking account
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:100]:
```suggestion
     * @param to Consensus address shares are delegated to
```

File: contracts/contracts/Subcall.sol [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+     * @param value native token amount (in wei)
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:101]:
```suggestion
     * @param value Native token amount (in wei)
```

File: contracts/contracts/Subcall.sol [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+            revert ConsensusDelegateError(status, string(data));
+        }
+    }
+
+    error ConsensusWithdrawError(uint64 status, string data);
+
+    /**
+     * Transfer from an account in this runtime to a consensus staking account.
+     *
+     * @param to consensus staking account which gets the tokens
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:120]:
```suggestion
     * @param to Consensus address which gets the tokens
```

File: contracts/contracts/Subcall.sol [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+     * @param value native token amount (in wei)
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:121]:
```suggestion
     * @param value Native token amount (in wei)
```

**Review by matevz — CHANGES_REQUESTED** [review]
Please rename `StakingPublicKey` to `StakingAddress`. Also let's use term "address" for destination address instead of "account" for delegation/undelegation/withdrawal.

**Comment by CedarMist** [comment]
> Please rename `StakingPublicKey` to `StakingAddress`. Also let's use term "address" for destination address instead of "account" for delegation/undelegation/withdrawal.

I'm not sure why I used public key, I even had a function called derive address from public key...

File: contracts/contracts/ConsensusUtils.sol [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+pragma solidity ^0.8.0;
+
+import {sha512_256, Sapphire} from "./Sapphire.sol";
+
+type StakingAddress is bytes21;
+
+type StakingSecretKey is bytes32;
+
+library ConsensusUtils {
+    string private constant ADDRESS_V0_CONTEXT_IDENTIFIER =
```
Comment (matevz) [inline contracts/contracts/ConsensusUtils.sol:12]:
```suggestion
    // https://github.com/oasisprotocol/oasis-core/blob/master/go/staking/api/address.go#L16
    string private constant ADDRESS_V0_CONTEXT_IDENTIFIER =
```

**Review by matevz — CHANGES_REQUESTED** [review]
Please document all external, public and internal functions and constants in Solidity code since that's what the docs will be generated of. Otherwise looks ok, will make another pass later.

File: contracts/contracts/ConsensusUtils.sol [hunk 15]
```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: Apache-2.0
+
+pragma solidity ^0.8.0;
+
+import {sha512_256, Sapphire} from "./Sapphire.sol";
+
+// 21 byte version-prefixed address (1 byte version, 20 bytes truncated digest )
```
Comment (matevz) [inline contracts/contracts/ConsensusUtils.sol:7]:
```suggestion
/// 21 byte version-prefixed address (1 byte version, 20 bytes truncated digest).
```

File: contracts/contracts/ConsensusUtils.sol [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+type StakingAddress is bytes21;
+
+/// 32 byte secret key
```
Comment (matevz) [inline contracts/contracts/ConsensusUtils.sol:10]:
```suggestion
/// 32 byte secret key.
```

File: contracts/contracts/Subcall.sol [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+    string private constant CONSENSUS_DELEGATE = "consensus.Delegate";
+    string private constant CONSENSUS_UNDELEGATE = "consensus.Undelegate";
+    string private constant CONSENSUS_WITHDRAW = "consensus.Withdraw";
+    string private constant ACCOUNTS_TRANSFER = "accounts.Transfer";
+
+    /// Address of the SUBCALL precompile
+    address internal constant SUBCALL =
+        0x0100000000000000000000000000000000000103;
+
+    /// Only raised if the underlying subcall precompile does not succeed
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:21]:
```suggestion
    /// Raised if the underlying subcall precompile does not succeed.
```

File: contracts/contracts/Subcall.sol [hunk 18]
```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: Apache-2.0
+
+pragma solidity ^0.8.0;
+
+import {StakingAddress, StakingSecretKey} from "./ConsensusUtils.sol";
+
+/**
+ * @title SDK Subcall wrappers
+ * @dev Interact with Oasis Runtime SDK modules from Sapphire
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:9]:
```suggestion
 * @dev Interact with Oasis Runtime SDK modules from Sapphire.
```

File: contracts/contracts/Subcall.sol [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+
+    /// Address of the SUBCALL precompile
+    address internal constant SUBCALL =
+        0x0100000000000000000000000000000000000103;
+
+    /// Only raised if the underlying subcall precompile does not succeed
+    error Subcall_Error();
+
+    /**
+     * Submit a native message to the Oasis runtime layer
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:25]:
```suggestion
     * Submit a native message to the Oasis runtime layer.
```

File: contracts/contracts/Subcall.sol [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+
+        if (status != 0) {
+            revert ConsensusUndelegateError(status, string(data));
+        }
+    }
+
+    error ConsensusDelegateError(uint64 status, string data);
+
+    /**
+     * Delegate native token to consensus level
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:110]:
```suggestion
     * Delegate native token to consensus level.
```

File: contracts/contracts/Subcall.sol [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+                hex"66",
+                "amount",
+                hex"8250",
+                value,
+                hex"40"
+            )
+        );
+    }
+
+    error ConsensusUndelegateError(uint64 status, string data);
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:78]: Let's move all contract-level errors above after the constants and store them in one place.

File: contracts/contracts/Subcall.sol [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+                shares
+            )
+        );
+
+        if (status != 0) {
+            revert ConsensusUndelegateError(status, string(data));
+        }
+    }
+
+    error ConsensusDelegateError(uint64 status, string data);
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:107]: Move above.

File: contracts/contracts/Subcall.sol [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+            to,
+            value
+        );
+
+        if (status != 0) {
+            revert ConsensusDelegateError(status, string(data));
+        }
+    }
+
+    error ConsensusWithdrawError(uint64 status, string data);
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:127]: Move above.

File: contracts/contracts/Subcall.sol [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+            to,
+            value
+        );
+
+        if (status != 0) {
+            revert ConsensusWithdrawError(status, string(data));
+        }
+    }
+
+    error AccountsTransferError(uint64 status, string data);
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:147]: Move above.

File: contracts/contracts/Subcall.sol [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+
+    /**
+     * Perform a transfer to another account
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:150]:
```suggestion
     * Perform a transfer to another account.
```

File: contracts/test/subcall.ts [hunk 26]
```diff
@@ -1,28 +1,184 @@
...
+    );
+    const computedPublicKey = ethers.utils.hexlify(
+      await oasis.staking.addressFromPublicKey(alice.public()),
+    );
+
+    expect(computedPublicKey).eq(ethers.utils.hexlify(newKeypair.publicKey));
+  });
+
+  it('accounts.Transfer', async () => {
+    // Ensure contract has an initial balance
```
Comment (matevz) [inline contracts/test/subcall.ts:95]:
```suggestion
    // Ensure contract has an initial balance.
```

Please add full-stops at the end of regular comments. Also below.

File: contracts/contracts/Subcall.sol [hunk 27]
```diff
@@ -0,0 +1,168 @@
...
+    function subcall(string memory method, bytes memory body)
+        internal
+        returns (uint64 status, bytes memory data)
+    {
+        (bool success, bytes memory tmp) = SUBCALL.call(
+            abi.encode(method, body)
+        );
+
+        if (!success) {
+            revert Subcall_Error();
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:43]: Is there a way to retrieve actual subcall error?

**Review by matevz — APPROVED** [review]
Some minor stylstic nits, otherwise LGTM.

**Pull request merged**